frontend/src/components/ui/chat/ChatBlock.js

```diff
@@ -1,41 +1,83 @@
-import { useState } from "react";
-
-import TextField from "@mui/material/TextField";
-import Divider from "@mui/material/Divider";
-
-import ChatList from "./ChatList";
-import Title from "./Title";
-import styles from "./ChatBlock.module.css";
-
-const DUMMY_DATA = [
-  {
-    senderId: "perborgen",
-    text: "who'll win!who'll win!who'll win!who'll win!who'll win!who'll win!who'll win!",
-  },
-  {
-    senderId: "janedoe",
-    text: "who'll win!",
-  },
-  {
-    senderId: "jimyy",
-    text: "who'll win!",
-  },
-];
-
-function ChatBlock() {
-  const[state] = useState({ chats: DUMMY_DATA });
+import { useState, useEffect, useContext } from 'react';
+
+import TextField from '@mui/material/TextField';
+import Divider from '@mui/material/Divider';
+
+import ChatList from './ChatList';
+import Title from './Title';
+import styles from './ChatBlock.module.css';
+
+import useInput from '../../../hooks/use-input';
+
+import UserContext from '../../../context/user-context';
+
+import isValidChat from '../../../validators/chat-validator';
+
+const ENTER_KEY_CODE = 13;
+
+function ChatBlock({ socket, roomId }) {
+  const userCtx = useContext(UserContext);
+  const [chats, setChats] = useState([]);
+
+  const {
+    value: chatMessage,
+    isValid: chatIsValid,
+    valueChangeHandler: chatChangeHandler,
+    reset: resetChat,
+  } = useInput(isValidChat);
+
+  useEffect(() => {
+    // Using useeffect hook to avoid rendering duplicate copies of chat received
+    // (Only render chat block component once after receive message)
+    socket.on('receive-chat', (senderUsername, chatMessage) => {
+      setChats((chats) => [...chats, { senderUsername, text: chatMessage }]);
+    });
+
+    // The listeners must be removed in the cleanup step,
+    // in order to prevent multiple event registrations.
+    return () => socket.off('receive-chat');
+  }, [socket]);
+
+  const sendChatHandler = (event) => {
+    if (event.keyCode !== ENTER_KEY_CODE) {
+      // If the key pressed is not ENTER key, do not proceed to send the message
+      return;
+    }
+
+    event.preventDefault();
+
+    if (!chatIsValid) {
+      // If the message is empty, do not proceed to send the message
+      resetChat();
+      return;
+    }
+
+    const senderUsername = userCtx.username;
+
+    socket.emit('send-chat', roomId, senderUsername, chatMessage);
+
+    setChats((chats) => [
+      ...chats,
+      { senderUsername: 'Me', text: chatMessage },
+    ]);
+
+    resetChat();
+  };
```

Comment:
```suggestion
  const sendChatHandler = (event) => {
    if (event.keyCode === ENTER_KEY_CODE) {
      event.preventDefault();
      
      if (!chatIsValid) {
        // If the message is empty, do not proceed to send the message
        resetChat();
        return;
      }
  
      const senderUsername = userCtx.username;
  
      socket.emit('send-chat', roomId, senderUsername, chatMessage);
  
      setChats((chats) => [
        ...chats,
        { senderUsername: 'Me', text: chatMessage },
      ]);
  
      resetChat();
    }
    // If the key pressed is not ENTER key, do not proceed to send the message
  };
```

Reply:
Whoa didn't know there's this feature in Github, amazing haha.

Yea your code is much cleaner! I shall adopt yours.